programs/marinade-finance/src/instructions/management/redelegate.rs

```diff
@@ -0,0 +1,368 @@
+        mut,
+        address = state.stake_system.stake_list.account,
+        constraint = stake_list.data.borrow().as_ref().get(0..8)
+            == Some(StakeSystem::DISCRIMINATOR)
+            @ MarinadeError::InvalidStakeListDiscriminator,
+    )]
+    pub stake_list: UncheckedAccount<'info>,
+    #[account(mut)]
+    pub stake_account: Box<Account<'info, StakeAccount>>,
+    /// CHECK: PDA
+    #[account(
+        seeds = [
+            &state.key().to_bytes(),
+            StakeSystem::STAKE_DEPOSIT_SEED
+        ],
+        bump = state.stake_system.stake_deposit_bump_seed
+    )]
+    pub stake_deposit_authority: UncheckedAccount<'info>,
+    // Readonly. For stake delta calculation
+    #[account(
+        seeds = [
+            &state.key().to_bytes(),
+            State::RESERVE_SEED
+        ],
+        bump = state.reserve_bump_seed
+    )]
+    pub reserve_pda: SystemAccount<'info>,
+
+    #[account(
+        init,
+        payer = split_stake_rent_payer,
+        space = std::mem::size_of::<StakeState>(),
+        owner = stake::program::ID,
+    )]
+    pub split_stake_account: Account<'info, StakeAccount>,
+    #[account(
+        mut,
+        owner = system_program::ID
+    )]
+    pub split_stake_rent_payer: Signer<'info>,
+
+    /// CHECK: compared to value stored in list
+    pub dest_validator_account: UncheckedAccount<'info>,
+    // new stake account to make the reDelegation
+    #[account(
+        init,
+        payer = split_stake_rent_payer,
+        space = std::mem::size_of::<StakeState>(),
+        owner = stake::program::ID,
+    )]
+    pub redelegate_stake_account: Account<'info, StakeAccount>,
+
+    pub clock: Sysvar<'info, Clock>,
+    /// CHECK: have no CPU budget to parse
+    pub stake_history: UncheckedAccount<'info>,
+    /// CHECK: have no CPU budget to parse
+    pub stake_config: UncheckedAccount<'info>,
+
+    pub system_program: Program<'info, System>,
+    pub stake_program: Program<'info, Stake>,
+}
+
+impl<'info> ReDelegate<'info> {
+    /// Notes on Solana re-delegation of stake account:
+    /// 1. In order to redelegate you need to create a new stake account
+    /// 2. The old stake account remains in the system until the end of the epoch, cooling down
+    /// 3. The new stake account exists at the same time, warming up
+    /// 4. It means that the same delegation.stake "exists" in two accounts at the same time, one cooling down, the other warming up
+    ///
+    /// What we do here:
+    /// 1. Check if we need to split the account (partial redelegate) or we are re-delegating the entire stake account
+    /// 2. if needed, split the account
+    /// 3. re-delegate the stake-account or the splitted account, into the new stake account + new validator
+    ///
+    pub fn process(
+        &mut self,
+        stake_index: u32,
+        validator_index: u32,
+        dest_validator_index: u32,
+        desired_redelegate_amount: u64,
+    ) -> Result<()> {
+        assert!(
```

Comment:
No assert for the input parameter!!! Never do this!!! Assert must protect from code bugs, not wrong inputs. Use require! from anchor to return the custom error code